src/test/java/BaseTest.java

```diff
@@ -1,11 +1,109 @@
 import io.github.bonigarcia.wdm.WebDriverManager;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
 import org.openqa.selenium.chrome.ChromeOptions;
-import org.testng.annotations.BeforeSuite;
-
-public class BaseTest {
+import org.openqa.selenium.edge.EdgeDriver;
+import org.openqa.selenium.edge.EdgeOptions;
+import org.openqa.selenium.firefox.FirefoxDriver;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.annotations.*;
 
+import java.time.Duration;
+public class BaseTest{
+    public static WebDriver driver;
+    public WebDriverWait wait;
+    public String url = "https://qa.koel.app/";
+    public Actions actions;
     @BeforeSuite
     static void setupClass() {
-        WebDriverManager.chromedriver().setup();
+
+        //WebDriverManager.chromedriver().setup();
+        WebDriverManager.firefoxdriver().setup();
+    }
+    @BeforeMethod
+    @Parameters({"url"})
+    public void launchBrowser(String url) {
+        //ChromeOptions options = new ChromeOptions();
+       // options.addArguments("--remote-allow-origins=*");
+        //driver = new ChromeDriver(options);
+        //driver = new FirefoxDriver();
+        driver = pickBrowser(System.getProperty("browser"));
+
+        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
+        driver.manage().window().maximize();
+        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
+        actions = new Actions(driver);
+        driver.get(url);
+    }
+
+    //return driver = new RemoteWebDriver(new URL("https://" +userName + ":" +accessKey + hubURL), capabilities);
+    @AfterMethod
+    public void closeBrowser() {
+        driver.quit();
+    }
+
+    public static WebDriver pickBrowser(String browser){
+        switch(browser){
+            case "firefox":
+                WebDriverManager.firefoxdriver().setup();
+                return driver = new FirefoxDriver();
+            case "MicrosoftEdge":
+                WebDriverManager.edgedriver().setup();
+                EdgeOptions edgeOptions = new EdgeOptions();
+                edgeOptions.addArguments("--remote-allow-origins=*");
+                return driver = new EdgeDriver();
+            default:
+                WebDriverManager.chromedriver().setup();
+                ChromeOptions chromeOptions = new ChromeOptions();
+                chromeOptions.addArguments("--remote-allow-origins=*");
+                return driver = new ChromeDriver(chromeOptions);
+
+
+        }
+    }
```

Comment:
Please finish your homework.

Hint 

_You can run Chrome browser  or Firefox using your local Grid:_


**Step 1** **You don't have files (selenium-server-4.12.1.jar) in your project A51-QA-Automation-Framework**

For example

**Add the**  **selenium-server-4.12.1.jar**  file to the project
Download Remote Selenium WebDriver (Grid)  https://www.selenium.dev/downloads/

![image](https://github.com/rhieya416/A49-QA-Automation-Framework/assets/104761482/ce97b866-2dca-4fe2-9383-fed5992b0c13)

![image](https://github.com/rhieya416/A49-QA-Automation-Framework/assets/104761482/b8e5132e-219f-4fa3-aa17-9a245f259463)


**Step 2**

To start the selenium server:** `java -jar selenium-server-4.12.1.jar standalone --selenium-manager true`

![image](https://github.com/Test-Pro-Bootcamp/A49-QA-Automation-Framework/assets/104761482/051e9832-d0a8-4d05-a1c3-a4b9b59e7b7c)

**Step 3**

Add an URL to your project, for example

![image](https://github.com/Test-Pro-Bootcamp/A49-QA-Automation-Framework/assets/104761482/a9d0805a-89e4-4e3d-b4b9-303c2903d019)

`String gridURL = " http://192.168.0.248:4444";`

```
 public WebDriver pickBrowser(String browser, Map<String, String> variantCapabilities) throws MalformedURLException {
        DesiredCapabilities caps = new DesiredCapabilities(); // Create a new DesiredCapabilities object
        String gridURL = "http://192.168.0.248:4444"; // Set the URL for the Selenium Grid

        if (variantCapabilities != null) {
            caps.merge(new DesiredCapabilities(variantCapabilities)); // Merge any variant capabilities into the DesiredCapabilities object
        }

        switch (browser) {
            case "firefox":
                WebDriverManager.firefoxdriver().setup(); // Set up the FirefoxDriver executable using WebDriverManager
                FirefoxOptions firefoxOptions = new FirefoxOptions(); // Create a new FirefoxOptions object
                firefoxOptions.addPreference("dom.webnotifications.enabled", false); // Disable web notifications
                return new FirefoxDriver(firefoxOptions.merge(caps)); // Launch Firefox with the specified options and capabilities

            case "MicrosoftEdge":
                WebDriverManager.edgedriver().setup(); // Set up the EdgeDriver executable using WebDriverManager
                EdgeOptions edgeOptions = new EdgeOptions(); // Create a new EdgeOptions object
                edgeOptions.addArguments("--remote-allow-origins=*"); // Allow remote origins
                return new EdgeDriver(edgeOptions.merge(caps)); // Launch Edge with the specified options and capabilities

            case "grid-edge":
                caps.setCapability("browserName", "MicrosoftEdge"); // Set the browser capability to Microsoft Edge
                return new RemoteWebDriver(URI.create(gridURL).toURL(), caps); // Launch a remote WebDriver instance with the specified capabilities

            case "grid-firefox":
                caps.setCapability("browserName", "firefox"); // Set the browser capability to Firefox
                return new RemoteWebDriver(URI.create(gridURL).toURL(), caps); // Launch a remote WebDriver instance with the specified capabilities

            case "grid-chrome":
                caps.setCapability("browserName", "chrome"); // Set the browser capability to Chrome
                return new RemoteWebDriver(URI.create(gridURL).toURL(), caps); // Launch a remote WebDriver instance with the specified capabilities

            default:
                WebDriverManager.chromedriver().setup(); // Set up the ChromeDriver executable using WebDriverManager
                ChromeOptions options = new ChromeOptions(); // Create a new ChromeOptions object
                options.addArguments("--disable-notifications", "--remote-allow-origins=*", "--incognito", "--start-maximized"); // Disable notifications, allow remote origins, launch in incognito mode, and start maximized
                options.setExperimentalOption("excludeSwitches", new String[]{"enable-automation"}); // Exclude the enable-automation switch
                return new ChromeDriver(options.merge(caps)); // Launch Chrome with the specified options and capabilities
        }
    }
```

**Step 4**


![image](https://github.com/rhieya416/A49-QA-Automation-Framework/assets/104761482/5ee2eff9-2231-478e-b7a7-199fb0a4e2f1)

![image](https://github.com/rhieya416/A49-QA-Automation-Framework/assets/104761482/120a6812-8217-459f-af51-55c780def394)

Run command in console: `gradle clean test -Dbrowser=grid-firefox`

![image](https://github.com/rhieya416/A49-QA-Automation-Framework/assets/104761482/837fff1d-dbd4-4919-96fc-04b6f5c651fa)


![image](https://github.com/rhieya416/A49-QA-Automation-Framework/assets/104761482/dfaba966-2f48-465c-8f5e-2da897e3222f)

**GRID MacOS ( video)**  https://drive.google.com/file/d/1Tkk_mve-lTeBd4-iSs83OWyXNu0ruCtB/view?usp=sharing

[SeleniumGrid_forMac.pdf](https://github.com/Test-Pro-Bootcamp/A33-QA-Automation-Framework/files/13388392/SeleniumGrid_forMac.pdf)

**Full code** https://drive.google.com/drive/folders/1DwIrvwx6lZZ5SBV72Ys4ibCmZXpgsy4p?usp=sharing


Hope this at least helps you know where you stand.

Reply:
Hi, my test is running when i use gradle clean test -Dbrowser=grid-firefox, but on server site is no session, also could you help me how to move grid server to intellij, i have installed to my device and run on terminal, 
<img width="1440" alt="Screenshot 2024-04-16 at 3 47 50 PM" src="https://github.com/Test-Pro-Bootcamp/A57-QA-Automation-Framework/assets/163935677/c60c76ce-400c-4741-8eeb-287c34a6d67c">
<img width="1440" alt="Screenshot 2024-04-16 at 3 48 00 PM" src="https://github.com/Test-Pro-Bootcamp/A57-QA-Automation-Framework/assets/163935677/d6b1838a-617c-42a6-9899-bf5b70675a92">

Reply:
> Hi, my test is running when i use gradle clean test -Dbrowser=grid-firefox, but on server site is no session, also could you help me how to move grid server to intellij, i have installed to my device and run on terminal, <img alt="Screenshot 2024-04-16 at 3 47 50 PM" width="1440" src="https://private-user-images.githubusercontent.com/163935677/322981262-c60c76ce-400c-4741-8eeb-287c34a6d67c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzMzA5OTEsIm5iZiI6MTcxMzMzMDY5MSwicGF0aCI6Ii8xNjM5MzU2NzcvMzIyOTgxMjYyLWM2MGM3NmNlLTQwMGMtNDc0MS04ZWViLTI4N2MzNGE2ZDY3Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxN1QwNTExMzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MDQ4YTJmNWIyYjY0OTNhZjM3MTBkYjgxYmJhNGQ1MzM5YTVhNzQyODVkMjA1ZDIxN2VmY2NmZmQ3OGUyMjI5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.WyZS-5scbV_f6IzX26I0OM6jeZF_qCo-SO4hjYSq8aI"> <img alt="Screenshot 2024-04-16 at 3 48 00 PM" width="1440" src="https://private-user-images.githubusercontent.com/163935677/322981310-d6b1838a-617c-42a6-9899-bf5b70675a92.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzMzA5OTEsIm5iZiI6MTcxMzMzMDY5MSwicGF0aCI6Ii8xNjM5MzU2NzcvMzIyOTgxMzEwLWQ2YjE4MzhhLTYxN2MtNDJhNi05ODk5LWJmNWI3MDY3NWE5Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxN1QwNTExMzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMWRhYzJkZTk1OTVjOTc0ZjFkOWFiZDE2NTE3YTc3OWY1ZDdiMWZiZmI4YmI5MjQ1MDg5MTI1MDU3MzFlZGI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.2eZOaFGIj_8R0r7QQQDZmxNXFH40CiX-BE-_lYy_xuk">

Hi

Run commands:

`java -jar selenium-server-standalone-4.16.1.jar -role node`

**if that doesn't work then try this**

1) `java -jar selenium-server-4.16.1.jar hub`

This command starts a Selenium Grid Hub using the JAR file "selenium-server-4.16.1.jar". The hub is a central point where tests are queued up and then distributed to available nodes (machines running browsers).

2) `java -Dwebdriver.chrome.driver="C:\MyWebDriver\chromedriver.exe" -Dwebdriver.gecko.driver="C:\MyWebDriver\geckodriver.exe" -jar selenium-server-4.16.1.jar node --port 5555`

This command starts a Selenium Grid Node using the same JAR file. The node is where the actual browser instances run tests.


- -Dwebdriver.chrome.driver="C:\MyWebDriver\chromedriver.exe" specifies the path to the Chrome WebDriver executable.
- -Dwebdriver.gecko.driver="C:\MyWebDriver\geckodriver.exe" specifies the path to the Gecko (Firefox) WebDriver executable.
- --port 5555 specifies the port number on which the node will listen for connections from the hub.


or 
https://medium.com/chaya-thilakumara/selenium-grid-chrome-browser-c2cb986e0f70 

https://reflect.run/articles/setting-up-selenium-grid-for-cross-browser-testing/